[scripts-audit] vcpkg.cmake

strega-nil opened the pull request:
See PR #16055

This also (attempts to) allow vcpkg.cmake to run with cmake 3.1; we had a `VERSION_GREATER_EQUAL` check, but that if keyword was added in cmake 3.7.

BillyONeal commented on scripts/buildsystems/vcpkg.cmake:
```diff
@@ -586,23 +578,56 @@ if(X_VCPKG_APPLOCAL_DEPS_INSTALL)
+this macro allows a function to get a variable with its arguments,
+escaped correctly so that list arguments are not passed as splats to the underlying function.
+#]===]
+macro(z_vcpkg_escape_argv OUT)
```
Does it matter that this is used before it is declared?

BillyONeal replied:
Also I don't know how to audit this for correctness so other reviewers pay close attention.

strega-nil replied:
1. It doesn't but 2. I still think we should move z_vcpkg_escape_argv up.

strega-nil committed:
[vcpkg.cmake] modify to get in line with the audit

See PR #16055

This also (attempts to) allow vcpkg.cmake to run with cmake 3.0; we had
a VERSION_GREATER_EQUAL check, but that if keyword was added in cmake
3.7.

Neumann-A commented on scripts/buildsystems/vcpkg.cmake:
```diff
@@ -589,20 +731,29 @@ endif()
     set(VCPKG_OVERRIDE_FIND_PACKAGE_NAME find_package)
 endif()
-macro(${VCPKG_OVERRIDE_FIND_PACKAGE_NAME} name)
+function("${VCPKG_OVERRIDE_FIND_PACKAGE_NAME}")
```
this potentially scopes the `find_package` call which is not intended

Neumann-A commented on scripts/buildsystems/vcpkg.cmake:
```diff
@@ -217,76 +213,77 @@ else()
     else()
-        set(_VCPKG_INSTALLED_DIR ${_VCPKG_ROOT_DIR}/installed)
+        set(Z_VCPKG_INSTALLED_DIR "${Z_VCPKG_ROOT_DIR}/installed")
     endif()
 
-    set(_VCPKG_INSTALLED_DIR ${_VCPKG_INSTALLED_DIR}
+    set(Z_VCPKG_INSTALLED_DIR "${Z_VCPKG_INSTALLED_DIR}"
```
changing _VCPKG_INSTALLED_DIR probably requires replacement in a lot of vcpkg-cmake-wrapper.cmake

Neumann-A commented on scripts/buildsystems/vcpkg.cmake:
```diff
@@ -589,20 +731,29 @@ endif()
-        include(${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/share/${_vcpkg_lowercase_name}/vcpkg-cmake-wrapper.cmake)
-    elseif("${name}" STREQUAL "Boost" AND EXISTS "${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/include/boost")
+    z_vcpkg_escape_argv(ESCAPED_ARGV)
+    set(TARGET_NAME "${ARGV0}")
+    string(TOLOWER "${TARGET_NAME}" LOWERCASE_TARGET_NAME)
```
this is not a `TARGET_NAME` this is a `PACKAGE_NAME`. The CMake docs always uses `<PackageName>` to describe the variable passed via `find_package`

Neumann-A commented:
I would really like to see some cleanup first instead of more things added to `vcpkg.cmake`.

strega-nil committed:
s/target_name/package_name in find_package

ras0219-msft reviewed (commented):
The pieces I've reviewed LGTM so far.

Still to review:
- New macro/functions
- `find_package()` override

strega-nil commented:
Depends on #16130

strega-nil committed:
depend on z_vcpkg_function_arguments

instead of a custom macro
also, as a drive-by, switch to foreach(X IN LISTS ...)

strega-nil committed:
split the utility macros/functions out

additionally, move the options and settings to the top of the file

strega-nil committed:
move more options and settings to the top

strega-nil committed:
fix name of file

strega-nil committed:
fix ras code comments

* remove z_vcpkg_utilities due to export
* add cmake_policy(PUSH|POP)
* add VCPKG_INSTALLED_DIR input variable
* add .cmakestamp to vcpkg_installed

strega-nil committed:
fix osx ifs to use quotes

also, remove spaces between if and (, to keep style consistent
(mostly for ease of grepping...)

strega-nil commented:
/azp run

strega-nil committed:
fix bug fixed by #16193

strega-nil committed:
move z_vcpkg_add_fatal_error up

strega-nil committed:
fix RANGE for reals

strega-nil commented:
/azp run

BillyONeal commented on scripts/buildsystems/vcpkg.cmake:
```diff
@@ -84,16 +99,215 @@ Install the dependencies listed in your manifest:
+            if(Z_VCPKG_BUILTIN_POWERSHELL_PATH)
+                set(Z_VCPKG_POWERSHELL_PATH "${Z_VCPKG_BUILTIN_POWERSHELL_PATH}" CACHE INTERNAL "The path to the PowerShell implementation to use.")
+            else()
+                message(WARNING "vcpkg: Could not find PowerShell; using static string 'powershell.exe'")
```
I know this was here before but doesn't `REQUIRED` on 289 make this else dead?

strega-nil replied:
Yes, it does. We should remove it in the future.

BillyONeal commented on scripts/buildsystems/vcpkg.cmake:
```diff
@@ -581,28 +784,35 @@ if(X_VCPKG_APPLOCAL_DEPS_INSTALL)
-        set(ARGS "${ARGV}")
-        include(${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/share/${_vcpkg_lowercase_name}/vcpkg-cmake-wrapper.cmake)
-    elseif("${name}" STREQUAL "Boost" AND EXISTS "${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/include/boost")
+    set(VCPKG_CMAKE_WRAPPER_PATH "${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/share/${LOWERCASE_PACKAGE_NAME}/vcpkg-cmake-wrapper.cmake")
```
`${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}` is repeated a lot, should we introduce a `Z_VCPKG_CURRENT_INSTALLED_DIR` ?

strega-nil replied:
Yeah probably, but we'll do that in the future.

BillyONeal commented on scripts/buildsystems/vcpkg.cmake:
```diff
@@ -614,79 +824,77 @@ macro(${VCPKG_OVERRIDE_FIND_PACKAGE_NAME} name)
-set(_UNUSED ${CMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY})
-set(_UNUSED ${CMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY})
-set(_UNUSED ${CMAKE_INSTALL_SYSTEM_RUNTIME_LIBS_SKIP})
+set(Z_VCPKG_UNUSED "${CMAKE_ERROR_ON_ABSOLUTE_INSTALL_DESTINATION}")
```
Stupid ?: Why are we scribbling over a variable with things then throwing it away?

strega-nil replied:
These are equivalent to casts to `void`.

BillyONeal replied:
I see ,thanks!

BillyONeal commented on scripts/buildsystems/vcpkg.cmake:
```diff
@@ -614,79 +824,77 @@ macro(${VCPKG_OVERRIDE_FIND_PACKAGE_NAME} name)
+set(Z_VCPKG_UNUSED "${CMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY}")
+set(Z_VCPKG_UNUSED "${CMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY}")
+set(Z_VCPKG_UNUSED "${CMAKE_INSTALL_SYSTEM_RUNTIME_LIBS_SKIP}")
 
```
Should there be an `unset(Z_VCPKG_UNUSED)` here?

strega-nil replied:
Not really, it's unimportant.

cenit commented:
@strega-nil Did you try the vcpkg.cmake script after the work on a downstream project?
It looks like it broke vcpkg totally. No packages can be found. I can reproduce it locally and it is being triggered right now in every ci pipeline of mine.
A message is also just appeared about that: https://github.com/microsoft/vcpkg/issues/15956#issuecomment-781859104
